### DISCUSSION TITLE
settings redesign/adjustments

### DISCUSSION THREAD

**Discussion by ghost** [mortii/anki-morphs#218] [Ideas]
============================================================================================
EDIT BY @mortii:

Todo list from this discussion:
- [x] Rename `Move new cards [...]` -> `Move new cards without unknown morphs to the end of the due queue`
- [x] Rename the `Skip` tab > `Card Handling`, and include these options:
    - [x] skip cards [...]
    - [x] move new cards without unknown morphs [...]
    - [x] shift new cards that are not the first [...]
    - [x] suspend new cards with only known morphs
- [x] Move to `General` tab:
    - [x] read files in "known morph" folder [...]
    - [x] morphs are considered known [...]
    - [x] U and A shows [...]
    - [x] Only show U in the toolbar
- [x] Add `restore default settings` button to the `note filter` and `extra field tabs`
- [x] Disable "delete" button when no row is selected
- [x]  top level checkbox in `extra fields` tab will toggle whether the child checkboxes are greyed out or not
- [x] disambiguate `skip cards with only known morphs` and `move new cards [...]` https://github.com/mortii/anki-morphs/discussions/191#discussioncomment-9179846
- [x] disable/gray out the spin boxes of any settings that haven't been selected
- [x] redo cancel/save/okay buttons (https://github.com/mortii/anki-morphs/discussions/218#discussioncomment-9109229)
    - [x] `OK`: apply and close
    - [x] `Cancel`: discard and close
    - [x] `Apply`: apply, don't close
    - [x] `x`: warning if any changes haven't been applied. 

============================================================================================

### Is your feature request related to a problem?

I have a lot of cards with the same vocabulary terms on them.  I have enabled "skip cards that have unknown morphs already seen today" in settings.  But as soon as the card is learning and not unknown, i am shown a lot of cards on the same day all with the same vocabulary item.  I would rather only get shown one of these per day so that i can see more new words.

### Describe the solution you would like

Morphman has two related settings to help with this (although neither are perfect).  The "skip cards with fresh vocabulary" option only shows unknown cards, not learning or known cards.  And the "skip card if focus morph was already seen today" option will do what i want.  It seems that in morphman the focus morph refers to cards at any not mature stage, whereas the equivalent option in ankimorphs works unexpectedly after the initial viewing of the cards.

I would like another setting similar to this one, "skip cards that have learning morphs already seen today", so that i will not keep being fed new sentences with learning words on them instead of new vocabulary.  Thank you

**Comment by mortii** [comment]
Sure, we could add that :)

Just for clarification: do you not have the `Skip cards with only known morphs`, nor the `Move new cards that only have known morphs to the end of the due queue` options activated?

Also, I just realized that the `Move new cards that only have known morphs to the end of the due queue` is labelled wrong, it should be `Move new cards that do not have unknown morphs to the end of the due queue`.

**Comment by ghost** [comment]
I have `Skip cards with only known morphs` activated, i thought i had `Move new cards that only have known morphs to the end of the due queue` activated but it looks like in fact it isn't for me.  Let me make sure it is enabled and try tomorrow to see whether it provides the behaviour i'm looking for.  This could be my mistake.

Thank you!

**Comment by aleksejrs** [comment]
> it should be
> `Move new cards that do not have unknown morphs to the end of the due queue`.

"without"
`Move new cards without unknown morphs to the end of the due queue`

**Comment by ghost** [comment]
Having ensured that the options are both enabled, i am getting the behaviour i wanted.

I wonder if it would be worth reorganising the settings interface to group these options, as they seem to go hand in hand.  I know right now the settings are organised by system rather than behaviour, so i don't want to create a new issue if that isn't something you're interested in, but on of the reasons i didn't make the connection before you mentioned it is because i saw the moving new cards behaviour as a part of the skipping unknown cards behaviour.

Thanks for the help, i'm really enjoying using ankimorphs!

**Comment by mortii** [comment]
> I wonder if it would be worth reorganising the settings interface to group these options, as they seem to go hand in hand.

Hmm, it would definitely be less hidden. The only reason to not do that is because the other skip options take effect without having to recalc, so that might be confusing.

The `suspend new cards with only known morphs`, and `shift new cards [...]` options are also somewhat related to skipping, do you think those should be moved too?

**Comment by ghost** [comment]
The `recalc` section of the ankimorph settings seems to have two main purposes for what happens on recalc: first, it changes the data of notes themselves, as with changing what is in the `am-unknowns` field, updating whether notes are known or not; the second is changing the notes within a deck, by changing the order.  These settings seem to me to relate more to skipping.  Perhaps i would keep the first kind in the `recalc` page, and move the others to the `skip` page.

It should be noted that as far as i can tell, skipping is not anki terminology, it is ankimorph terminology.  Perhaps if we used anki terminology, the separations might make more sense?  I mean, looking at the deck option settings for anki, there are categories named `display order` and `burying`.  To me, these are quite nice titles, as long as you understand what burying is.  So it might make sense to update the `skip` section of ankimorphs settings to be called something like `display order`, and use the term "bury" to describe operations relating to skipping there.  Then the options without this term we will know are not making use of this built in anki feature.  Perhaps an icon or warning "(requires recalc)" could be added to make it clearer.

I also tend to recalc anyway after changing the settings, because of the "recalc to avoid unexpected behaviour" popup that appears indiscriminately whenever you save the settings, regardless of whether you change a setting which actually requires a recalc or not.  I don't know if i'm alone in this.

These are just some preliminary thoughts, it would probably be helpful to hear other people's opinions on the settings as well.  In fact when i started using the addon, i was mostly impressed with how comprehensive the documentation was, especially compared to morphman, and perhaps the easiest short term solution would just be to add a note on one of the `skip` setup pages to also check the related options on the `recalc` page, as i think this is probably a commonly desired behaviour.

**Comment by mortii** [comment]
@iwouldrathernotusegithub awesome feedback, thank you! I converted this to a discussion since the scope increased, and the reply feature is better. 

Adding/renaming tabs might give a better overview of the settings. I'm thinking something like this:
1. Rename the `Skip` tab > `Card Handling`, and include these options:
    1. skip cards [...]
    2. move new cards without unknown morphs [...]
    3. shift new cards that are not the first [...]
    4. suspend new cards with only known morphs
2. Add a `Known Morphs` tab with these options:
    1. read files in "known morph" folder [...]
    2. morphs are considered known [...]
    3. U and A shows [...]

thoughts?

**Comment by ghost** [comment]
Thanks!  Those changes seem really great.  I think that would make things much easier to navigate.

As this is now slightly larger in scope, i will bring up a point in the `extra fields` tab: i find the checkmarks next to the cards slightly confusing.  For me, it looks like this:

```
[x] listening
    [x] am-unknowns
    [x] am-unknowns-count
    [x] am-highlighted
    [x] am-score
```

If i don't want these extra fields, i can uncheck the checkbox next to `listening` and all the checkboxes below will become unchecked; this is expected.  But if don't want the score to be added, and uncheck the checkbox next to `am-score`, it looks like this:

```
[ ] listening
    [x] am-unknowns
    [x] am-unknowns-count
    [x] am-highlighted
    [ ] am-score
```

Notice that the top level checkbox has also become unchecked.  I think this is confusing: i think that having the top level checkbox unchecked reads that the below fields will not be added to the cards, regardless of whether they are checked or not.  It could also mean that they *will* be added, but that because they are not all checked the top level is also not.  I assume the latter is the case here, and that this is a limitation of the qtreewidget, but i am not familiar enough with qt to say for sure.  If that's the case, i guess there's nothing that can be done, and i haven't deleted these fields from my cards to experiment but i thought i would raise it anyway.

Another suggestion that is purely cosmetic would be an option in the `known morphs` tab along the lines of `only show unique morphs stat in toolbar`, because as you note in the documentation, for isolating languages the numbers are always the same.  I had a look at the code and right now the toolbar buttons seem to be hardcoded, so perhaps it's too niche a feature to prioritise but in a redesign thread i thought i'd bring it up.

**Reply by mortii** [reply]
> Notice that the top level checkbox has also become unchecked. I think this is confusing

I can see how that could be confusing. I actually implemented this logic myself as a quick way to select/un-select all the fields, it's not a built in feature in qt:
https://github.com/mortii/anki-morphs/blob/d5f82a824d7559f40620da424fc29430e23d16a9/ankimorphs/settings_dialog.py#L767-L802

What would you like to see instead of this current approach?

> Another suggestion that is purely cosmetic would be an option in the `known morphs` tab along the lines of `only show unique morphs stat in toolbar`

So you want an option to only have the `U: xxx` in the toolbar and not include the `A: xxx`? If that's the case then that would be super easy to implement, so we could do that :+1:

**Comment by ghost** [comment]
> What would you like to see instead of this current approach?

Hm, i'm not sure what i'd want to see instead.  For me it's mainly that the top level checkbox reads as controlling whether the child checkboxes will be modified or not, rather than as a toggle for them.  What i would probably choose is the top level checkbox will toggle whether the child checkboxes are greyed out or not, and they are all checked by default.  So on checking the `listening` checkbox in my example, they all become enabled, and unchecking it they all become disabled which gives you the convenience you liked, but then checking or unchecking the children doesn't change whether the top level is checked.  That is, when it is possible for you to toggle the children, the parent is always checked.  Does that makes sense?

> So you want an option to only have the U: xxx in the toolbar and not include the A: xxx?

Yes, exactly!

I also noticed that the `Note filters` and `extra fields` tabs don't have a `restore default ______ settings` button at the bottom like the other tabs.  I'm not sure if this is intentional, because for these two the default values are not really helpful as they must be changed by the user to be functional or not but i thought i'd bring it up in any case.

I feel a little like i'm clutching at straws here and i'm not contributing any code so please don't feel obliged to work on any of these things in a timely manner; the addon is already really really good :)

**Reply by mortii** [reply]
> What i would probably choose is the top level checkbox will toggle whether the child checkboxes are greyed out or not, and they are all checked by default.

That's very good, we should do that instead.

> I also noticed that the Note filters and extra fields tabs don't have a restore default ______ settings button at the bottom like the other tabs.

Yes, that was intentional, but the reasons for it are probably not applicable anymore, so it might be a good idea to add a restore button to those tabs too.

> I feel a little like i'm clutching at straws here and i'm not contributing any code so please don't feel obliged to work on any of these things in a timely manner; the addon is already really really good :)

No worries. As for the timeline, I probably won't get around to implementing these things for at least a month.

**Comment by mortii** [comment]
I edited the top post with a todo list to make it easier to track.

I also want to disable/gray out the spin boxes of any settings that haven't been selected, e.g:
![image](https://github.com/mortii/anki-morphs/assets/15674619/3a409334-0ed1-408d-9088-7169fa71e5e7)

Btw, do you have any opinions about the "cancel" button @iwouldrathernotusegithub? @aleksejrs mentioned some issues he has with it #212.

**Comment by ghost** [comment]
>do you have any opinions about the "cancel" button

I'm going to put my thoughts here rather than the other issue, but if you think it's better to have them there, i can move them over.

Usually i close the settings with the window manager close button, or press escape, but i agree it's not a great design.  I've added a screenshot of how my kde settings looks, and i think how lots of settings dialogues in windows work too which i think does all of what you want quite well:

![image](https://github.com/mortii/anki-morphs/assets/147626539/8291e3ce-44f5-4a70-a4b7-25eaa76b7aaf)

You can see there are three buttons: `okay`, `apply`, and `cancel`.  `apply` is greyed out because i haven't changed any of the settings.  They work like this:

- `okay` will always close the window and save the changes made.
- `apply` will be greyed out until you change a setting, then clicking will save changes and become greyed out again, but keep the window open.
- `cancel` will always close the window, not saving changes

If i click cancel, it does not prompt me to save changes.  To me this makes sense, because the word cancel has no such implication.   But if i try and close the window  with the window manager close button, i get a warning:

![image](https://github.com/mortii/anki-morphs/assets/147626539/459b4ae5-81a2-4360-8ffd-53cd6a68f3a0)

Again we have three buttons: `save`, `discard`, and `cancel`.  Here, `cancel` will take you back to the settings window, keeping it open, and neither save nor discard the changes.  `save` will save and close the window, and `discard` will discard the changes and close the window.

I think this scheme works well, and also has the advantage of using well established conventions, so that's what i'd suggest :)

**Reply by ghost** [reply]
To clarify, by "window manager close button" i mean the one in the title bar, which for me is in the top right

**Reply by mortii** [reply]
My god, that is perfect, thank you so much!

**Comment by mortii** [comment]
I'm adding a "general" tab for options that don't fit into any one category:

![Screenshot from 2024-06-14 10-18-44](https://github.com/mortii/anki-morphs/assets/15674619/dae3890e-a91e-491c-901d-c7c2dcfd0eee)

**Comment by mortii** [comment]
Included in the v3 update: https://github.com/mortii/anki-morphs/releases/tag/v3.0.0

Thanks!

**Reply by ghost** [reply]
Oh, i only just saw this.  Thanks for all your work on ankimorphs <3
